File: zikzak_inappwebview_android/android/src/main/java/wtf/zikzak/zikzak_inappwebview_android/security/CertificatePinningManager.java
package wtf.zikzak.zikzak_inappwebview_android.security;

import android.util.Base64;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.security.MessageDigest;
import java.security.cert.Certificate;
import java.security.cert.CertificateEncodingException;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.net.ssl.SSLSession;
import javax.net.ssl.X509TrustManager;

/**
 * Manages SSL/TLS certificate pinning to prevent MITM attacks
 * Implements public key pinning using SHA-256 hashes
 */
public class CertificatePinningManager {

    private static final String TAG = "CertificatePinning";

    /**
     * Certificate pinning configuration for a specific host
     */
    public static class PinConfiguration {
        private final String hostname;
        private final List<String> publicKeyHashes; // Base64 encoded SHA256 hashes
        private final boolean includeSubdomains;
        private final boolean allowExpiredCertificates;

        public PinConfiguration(
            @NonNull String hostname,
            @NonNull List<String> publicKeyHashes,
            boolean includeSubdomains,
            boolean allowExpiredCertificates
        ) {
            this.hostname = hostname;
            this.publicKeyHashes = new ArrayList<>(publicKeyHashes);
            this.includeSubdomains = includeSubdomains;
            this.allowExpiredCertificates = allowExpiredCertificates;
        }

        public PinConfiguration(@NonNull String hostname, @NonNull List<String> publicKeyHashes) {
            this(hostname, publicKeyHashes, false, false);
        }

        public String getHostname() {
            return hostname;
        }

        public List<String> getPublicKeyHashes() {
            return new ArrayList<>(publicKeyHashes);
        }

        public boolean includesSubdomains() {
            return includeSubdomains;
        }

        public boolean allowsExpiredCertificates() {
            return allowExpiredCertificates;
        }
    }

    private final Map<String, PinConfiguration> pinConfigurations = new HashMap<>();
    private boolean enabled = false;

    public CertificatePinningManager() {}

    /**
     * Enable certificate pinning
     */
    public void enable() {
        enabled = true;
        Log.d(TAG, "Certificate pinning enabled");
    }

    /**
     * Disable certificate pinning
     */
    public void disable() {
        enabled = false;
        Log.d(TAG, "Certificate pinning disabled");
    }

    /**
     * Check if certificate pinning is enabled
     */
    public boolean isEnabled() {
        return enabled;
    }

    /**
     * Add a pin configuration for a hostname
     */
    public void addPinConfiguration(@NonNull PinConfiguration config) {
        pinConfigurations.put(config.getHostname(), config);
        Log.d(TAG, "Added pin configuration for: " + config.getHostname());
    }

    /**
     * Remove pin configuration for a hostname
     */
    public void removePinConfiguration(@NonNull String hostname) {
        pinConfigurations.remove(hostname);
        Log.d(TAG, "Removed pin configuration for: " + hostname);
    }

    /**
     * Clear all pin configurations
     */
    public void clearAllPinConfigurations() {
        pinConfigurations.clear();
        Log.d(TAG, "Cleared all pin configurations");
    }

    /**
     * Validate certificate chain against pinned certificates
     *
     * @param chain The certificate chain to validate
     * @param hostname The hostname being connected to
     * @return true if validation passes, false otherwise
     */
    public boolean validateCertificateChain(
        @Nullable Certificate[] chain,
        @NonNull String hostname
    ) {
        if (!enabled) {
            // Certificate pinning disabled, allow all
            return true;
        }

        if (chain == null || chain.length == 0) {
            Log.w(TAG, "Empty certificate chain for: " + hostname);
            return false;
        }

        // Check if we have a pin configuration for this host
        PinConfiguration config = getPinConfiguration(hostname);
        if (config == null) {
            // No pinning configured for this host, allow
            return true;
        }

        // Extract public key hashes from the certificate chain
        List<String> serverKeyHashes = new ArrayList<>();
        for (Certificate cert : chain) {
            if (cert instanceof X509Certificate) {
                try {
                    byte[] publicKey = cert.getPublicKey().getEncoded();
                    String hash = sha256Base64(publicKey);
                    serverKeyHashes.add(hash);
                } catch (Exception e) {
                    Log.e(TAG, "Failed to extract public key: " + e.getMessage());
                }
            }
        }

        if (serverKeyHashes.isEmpty()) {
            Log.w(TAG, "Failed to extract any public keys from certificate chain");
            return false;
        }

        // Check if any public key matches our pins
        for (String serverHash : serverKeyHashes) {
            if (config.getPublicKeyHashes().contains(serverHash)) {
                Log.d(TAG, "Certificate pinning validation passed for: " + hostname);
                return true;
            }
        }

        Log.w(TAG, "Certificate pinning validation FAILED for: " + hostname);
        Log.w(TAG, "Expected one of: " + config.getPublicKeyHashes());
        Log.w(TAG, "Got: " + serverKeyHashes);
        return false;
    }

    /**
     * Validate SSL session against pinned certificates
     *
     * @param sslSession The SSL session to validate
     * @param hostname The hostname being connected to
     * @return true if validation passes, false otherwise
     */
    public boolean validateSSLSession(
        @Nullable SSLSession sslSession,
        @NonNull String hostname
    ) {
        if (sslSession == null) {
            return false;
        }

        try {
            Certificate[] chain = sslSession.getPeerCertificates();
            return validateCertificateChain(chain, hostname);
        } catch (Exception e) {
            Log.e(TAG, "Failed to validate SSL session: " + e.getMessage());
            return false;
        }
    }

    /**
     * Get pin configuration for a hostname
     * Supports subdomain matching if configured
     */
    @Nullable
    private PinConfiguration getPinConfiguration(@NonNull String hostname) {
        // Direct match
        PinConfiguration config = pinConfigurations.get(hostname);
        if (config != null) {
            return config;
        }

        // Check for parent domain matches with includeSubdomains
        String[] parts = hostname.split("\\.");
        for (int i = 1; i < parts.length; i++) {
            String parentDomain = String.join(".", java.util.Arrays.copyOfRange(parts, i, parts.length));
            config = pinConfigurations.get(parentDomain);
            if (config != null && config.includesSubdomains()) {
                Log.d(TAG, "Using parent domain pin config for " + hostname + " from " + parentDomain);
                return config;
            }
        }

        return null;
    }

    /**
     * Calculate SHA-256 hash of data and return as Base64 string
     */
    @NonNull
    private String sha256Base64(@NonNull byte[] data) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(data);
            return Base64.encodeToString(hash, Base64.NO_WRAP);
        } catch (Exception e) {
            Log.e(TAG, "SHA-256 calculation failed: " + e.getMessage());
            return "";
        }
    }

    /**
     * Generate SHA-256 hash for a certificate's public key
     * Useful for generating pins
     *
     * @param certificate The certificate
     * @return Base64 encoded SHA-256 hash of the public key
     */
    @Nullable
    public static String generatePublicKeyHash(@NonNull X509Certificate certificate) {
        try {
            byte[] publicKey = certificate.getPublicKey().getEncoded();
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(publicKey);
            return Base64.encodeToString(hash, Base64.NO_WRAP);
        } catch (Exception e) {
            Log.e(TAG, "Failed to generate public key hash: " + e.getMessage());
            return null;
        }
    }

    /**
     * Get current pin configurations (for debugging)
     */
    @NonNull
    public Map<String, PinConfiguration> getPinConfigurations() {
        return new HashMap<>(pinConfigurations);
    }
}

File: zikzak_inappwebview_ios/ios/zikzak_inappwebview_ios/Sources/zikzak_inappwebview_ios/Security/CertificatePinningManager.swift
//
//  CertificatePinningManager.swift
//  zikzak_inappwebview_ios
//
//  Created by ZikZak Team on 2025-11-05.
//

import Foundation
import Security
import CommonCrypto

/// Manages SSL/TLS certificate pinning to prevent MITM attacks
public class CertificatePinningManager {

    /// Certificate pinning configuration for a specific host
    public struct PinConfiguration {
        let hostname: String
        let publicKeyHashes: [String] // Base64 encoded SHA256 hashes
        let includeSubdomains: Bool
        let allowExpiredCertificates: Bool

        public init(hostname: String,
                   publicKeyHashes: [String],
                   includeSubdomains: Bool = false,
                   allowExpiredCertificates: Bool = false) {
            self.hostname = hostname
            self.publicKeyHashes = publicKeyHashes
            self.includeSubdomains = includeSubdomains
            self.allowExpiredCertificates = allowExpiredCertificates
        }
    }

    private var pinConfigurations: [String: PinConfiguration] = [:]
    private var enabled: Bool = false

    public init() {}

    /// Enable certificate pinning
    public func enable() {
        enabled = true
    }

    /// Disable certificate pinning
    public func disable() {
        enabled = false
    }

    /// Add a pin configuration for a hostname
    public func addPinConfiguration(_ config: PinConfiguration) {
        pinConfigurations[config.hostname] = config
    }

    /// Remove pin configuration for a hostname
    public func removePinConfiguration(hostname: String) {
        pinConfigurations.removeValue(forKey: hostname)
    }

    /// Clear all pin configurations
    public func clearAllPinConfigurations() {
        pinConfigurations.removeAll()
    }

    /// Validate server trust against pinned certificates
    /// - Parameters:
    ///   - serverTrust: The server trust to validate
    ///   - hostname: The hostname being connected to
    /// - Returns: true if validation passes, false otherwise
    public func validateServerTrust(_ serverTrust: SecTrust, hostname: String) -> Bool {
        guard enabled else {
            // Certificate pinning disabled, allow all
            return true
        }

        // Check if we have a pin configuration for this host
        guard let config = getPinConfiguration(for: hostname) else {
            // No pinning configured for this host, allow
            return true
        }

        // Validate the certificate chain
        var error: CFError?
        guard SecTrustEvaluateWithError(serverTrust, &error) else {
            if let error = error {
                print("[CertificatePinning] Trust evaluation failed: \(error)")
            }
            return false
        }

        // Extract public keys from the certificate chain
        guard let serverPublicKeys = extractPublicKeys(from: serverTrust) else {
            print("[CertificatePinning] Failed to extract public keys")
            return false
        }

        // Check if any public key matches our pins
        for serverKey in serverPublicKeys {
            let serverKeyHash = sha256(data: serverKey)
            let serverKeyHashBase64 = serverKeyHash.base64EncodedString()

            if config.publicKeyHashes.contains(serverKeyHashBase64) {
                print("[CertificatePinning] ✅ Certificate pinning validation passed for \(hostname)")
                return true
            }
        }

        print("[CertificatePinning] ❌ Certificate pinning validation FAILED for \(hostname)")
        print("[CertificatePinning] None of the server's public keys match the pinned hashes")
        return false
    }

    /// Get pin configuration for a hostname (supports subdomains)
    private func getPinConfiguration(for hostname: String) -> PinConfiguration? {
        // Exact match
        if let config = pinConfigurations[hostname] {
            return config
        }

        // Check for subdomain matches
        for (configHost, config) in pinConfigurations {
            if config.includeSubdomains && hostname.hasSuffix("." + configHost) {
                return config
            }
        }

        return nil
    }

    /// Extract public keys from the certificate chain
    private func extractPublicKeys(from serverTrust: SecTrust) -> [Data]? {
        var publicKeys: [Data] = []

        let certificateCount = SecTrustGetCertificateCount(serverTrust)
        guard certificateCount > 0 else {
            return nil
        }

        for index in 0..<certificateCount {
            guard let certificate = SecTrustGetCertificateAtIndex(serverTrust, index) else {
                continue
            }

            guard let publicKey = extractPublicKey(from: certificate) else {
                continue
            }

            publicKeys.append(publicKey)
        }

        return publicKeys.isEmpty ? nil : publicKeys
    }

    /// Extract public key from a certificate
    private func extractPublicKey(from certificate: SecCertificate) -> Data? {
        // Get the public key from the certificate
        guard let publicKey = SecCertificateCopyKey(certificate) else {
            return nil
        }

        // Get the external representation of the public key
        var error: Unmanaged<CFError>?
        guard let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, &error) as Data? else {
            if let error = error?.takeRetainedValue() {
                print("[CertificatePinning] Failed to get public key data: \(error)")
            }
            return nil
        }

        return publicKeyData
    }

    /// Compute SHA256 hash of data
    private func sha256(data: Data) -> Data {
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(data.count), &hash)
        }
        return Data(hash)
    }

    /// Helper to create a pin configuration from a certificate file
    public static func createPinConfiguration(
        hostname: String,
        certificateData: Data,
        includeSubdomains: Bool = false
    ) -> PinConfiguration? {
        guard let certificate = SecCertificateCreateWithData(nil, certificateData as CFData) else {
            print("[CertificatePinning] Failed to create certificate from data")
            return nil
        }

        guard let publicKey = SecCertificateCopyKey(certificate) else {
            print("[CertificatePinning] Failed to extract public key from certificate")
            return nil
        }

        var error: Unmanaged<CFError>?
        guard let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, &error) as Data? else {
            print("[CertificatePinning] Failed to get public key data")
            return nil
        }

        // Compute SHA256 hash
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        publicKeyData.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(publicKeyData.count), &hash)
        }
        let hashData = Data(hash)
        let hashBase64 = hashData.base64EncodedString()

        return PinConfiguration(
            hostname: hostname,
            publicKeyHashes: [hashBase64],
            includeSubdomains: includeSubdomains
        )
    }
}
